firestore/src/android/firestore_android.cc

```diff
@@ -515,6 +542,59 @@ void FirestoreInternal::ClearListeners() {
   listener_registrations_.clear();
 }
 
+Future<LoadBundleTaskProgress> FirestoreInternal::LoadBundle(
+    const std::string& bundle) {
+  Env env = GetEnv();
+  Local<LoadBundleTaskInternal> task = CreateLoadBundleTask(env, obj_, bundle);
+  return promises_->NewFuture<LoadBundleTaskProgress>(env, AsyncFn::kLoadBundle,
+                                                      task);
+}
+
+Future<LoadBundleTaskProgress> FirestoreInternal::LoadBundle(
+    const std::string& bundle,
+    std::function<void(const LoadBundleTaskProgress&)> progress_callback) {
+  Env env = GetEnv();
+  Local<LoadBundleTaskInternal> task = CreateLoadBundleTask(env, obj_, bundle);
+
+  LambdaEventListener<LoadBundleTaskProgress> listener(
+      // TODO(C++14): Move the callback into the lambda.
+      [progress_callback](const LoadBundleTaskProgress& p, Error,
+                          const std::string&) {
+        // Error is always `kErrorOk` and error message is always empty for
+        // progress listeners.
+        progress_callback(p);
+      });
+
+  EventListener<LoadBundleTaskProgress>* listener_ptr = nullptr;
+  // TODO(b/187420421): The listener is owned by the firestore instance, longer
+  // than ideal. This is to support the unlikely case when user try to delete
+  // Firestore instance in the listener. Once the referred bug is fixed, it can
+  // be managed through a `shared_ptr`.
+  {
+    MutexLock lock(bundle_listeners_mutex_);
+    bundle_listeners_.push_back(std::move(listener));
+    listener_ptr = &(bundle_listeners_.back());
```

Comment:
Pointers to vector elements are not stable because they can be relocated with a future call to push_back(). This means that `listener_ptr` could become invalid in the future, which would cause undefined behavior in the progress updates.

Two possible solutions are (a) replace std::vector with a data structure that maintains pointer stability such as std::list or (b) change the vector to store pointers directly and allocate the objects with an explicit use of the `new` operator.

Reply:
Great catch, Denver! Sorry, I missed that the pointer gets stored for later usage.